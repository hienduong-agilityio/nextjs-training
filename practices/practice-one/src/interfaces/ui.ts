export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
