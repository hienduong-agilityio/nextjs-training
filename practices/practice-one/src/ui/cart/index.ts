export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
