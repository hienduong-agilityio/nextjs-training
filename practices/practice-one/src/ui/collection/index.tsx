export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
