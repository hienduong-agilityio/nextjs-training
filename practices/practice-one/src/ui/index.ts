export * from './products/index';
export * from './collection/index';
