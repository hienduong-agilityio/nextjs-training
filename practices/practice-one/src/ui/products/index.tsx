export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
