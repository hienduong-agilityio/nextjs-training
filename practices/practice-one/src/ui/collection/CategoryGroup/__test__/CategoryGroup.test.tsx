import { render, screen, fireEvent } from '@testing-library/react';
import { usePathname, useSearchParams } from 'next/navigation';
import { CategoryGroup } from '@/ui';

// Mock React hooks
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useOptimistic: jest.fn(),
}));

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

const mockUsePathname = usePathname as jest.Mock;
const mockUseSearchParams = useSearchParams as jest.Mock;
const mockUseOptimistic = jest.requireMock('react').useOptimistic;

// Mock data
const mockTitle = 'Categories';
const mockItems = [
  { name: 'All', count: 10 },
  { name: 'Bags', count: 5 },
  { name: 'Sneakers', count: 8 },
];

const setupMocks = (
  pathname = '/products',
  searchParams = new URLSearchParams(),
  optimisticCategory = 'All',
  setOptimisticCategory = jest.fn(),
) => {
  mockUsePathname.mockReturnValue(pathname);
  mockUseSearchParams.mockReturnValue(searchParams);
  mockUseOptimistic.mockReturnValue([
    optimisticCategory,
    setOptimisticCategory,
  ]);
  return setOptimisticCategory;
};

const renderComponent = (items = mockItems) => {
  render(<CategoryGroup title={mockTitle} items={items} />);
};

const expectCategoryStyles = (category: string, isActive: boolean) => {
  const categoryElement = screen.getByText(category);
  const activeClass = 'text-primary-200 bg-secondary-300';

  if (isActive) {
    expect(categoryElement).toHaveClass(activeClass);
  } else {
    expect(categoryElement).not.toHaveClass(activeClass);
  }
};

describe('CategoryGroup Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with title and category list', () => {
    setupMocks();
    renderComponent();

    expect(screen.getByText(mockTitle)).toBeInTheDocument();
    mockItems.forEach(({ name }) => {
      expect(screen.getByText(name)).toBeInTheDocument();
    });
  });

  it('applies correct styles for active and inactive categories', () => {
    setupMocks('/products', new URLSearchParams('category=Bags'), 'Bags');
    renderComponent();

    expectCategoryStyles('Bags', true);
    expectCategoryStyles('Sneakers', false);
    expectCategoryStyles('All', false);
  });

  it('updates the URL with query parameters on category click', () => {
    const mockSetOptimisticCategory = setupMocks();
    renderComponent();

    fireEvent.click(screen.getByText('Bags'));
    expect(mockSetOptimisticCategory).toHaveBeenCalledWith('Bags');
  });

  it('handles the ALL category properly when no category is selected', () => {
    setupMocks();
    renderComponent();

    expectCategoryStyles('All', true);
  });

  it('renders correctly with an empty items array', () => {
    setupMocks();
    renderComponent([]);

    expect(screen.getByText(mockTitle)).toBeInTheDocument();
    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });
});
