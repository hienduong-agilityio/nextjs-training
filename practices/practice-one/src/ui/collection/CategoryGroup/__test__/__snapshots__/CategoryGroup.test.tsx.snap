// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryGroup Component applies correct styles for active and inactive categories 1`] = `
<div>
  <div
    class="mb-6 bg-secondary-100 p-4 rounded"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Categories
    </h3>
    <ul
      class="space-y-2 text-sm text-gray-700"
    >
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?"
        >
          All
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=beauty"
        >
          Beauty
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=fragrances"
        >
          Fragrances
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded hover:bg-secondary-300 text-primary-200 bg-secondary-300"
          href="/products?category=furniture"
        >
          Furniture
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=groceries"
        >
          Groceries
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`CategoryGroup Component handles the ALL category properly when no category is selected 1`] = `
<div>
  <div
    class="mb-6 bg-secondary-100 p-4 rounded"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Categories
    </h3>
    <ul
      class="space-y-2 text-sm text-gray-700"
    >
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded hover:bg-secondary-300 text-primary-200 bg-secondary-300"
          href="/products?"
        >
          All
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=beauty"
        >
          Beauty
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=fragrances"
        >
          Fragrances
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=furniture"
        >
          Furniture
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=groceries"
        >
          Groceries
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`CategoryGroup Component renders correctly with an empty items array 1`] = `
<div>
  <div
    class="mb-6 bg-secondary-100 p-4 rounded"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Categories
    </h3>
    <ul
      class="space-y-2 text-sm text-gray-700"
    />
  </div>
</div>
`;

exports[`CategoryGroup Component renders correctly with title and category list 1`] = `
<div>
  <div
    class="mb-6 bg-secondary-100 p-4 rounded"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Categories
    </h3>
    <ul
      class="space-y-2 text-sm text-gray-700"
    >
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded hover:bg-secondary-300 text-primary-200 bg-secondary-300"
          href="/products?"
        >
          All
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=beauty"
        >
          Beauty
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=fragrances"
        >
          Fragrances
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=furniture"
        >
          Furniture
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=groceries"
        >
          Groceries
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`CategoryGroup Component updates the URL with query parameters on category click 1`] = `
<div>
  <div
    class="mb-6 bg-secondary-100 p-4 rounded"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Categories
    </h3>
    <ul
      class="space-y-2 text-sm text-gray-700"
    >
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded hover:bg-secondary-300 text-primary-200 bg-secondary-300"
          href="/products?"
        >
          All
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=beauty"
        >
          Beauty
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=fragrances"
        >
          Fragrances
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=furniture"
        >
          Furniture
        </a>
      </li>
      <li>
        <a
          class="w-full flex justify-start font-semibold py-2 rounded bg-secondary-100 hover:bg-secondary-300 text-gray-800"
          href="/products?category=groceries"
        >
          Groceries
        </a>
      </li>
    </ul>
  </div>
</div>
`;
